### PR TITLE
[codegen/*] Improve range type binding + codegen.

### PR DESCRIPTION
**Pull request by pgavlin**
- Determine variable types for ranged resources by typechecking an
  equivalent expression
- Detect top-level await in NodeJS and generate an async main
- Fix `pulumi.all` generation for NodeJS

### CODE REVIEW

File: pkg/codegen/internal/test/testdata/aws-eks.pp.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        }));
+    }
+    const rta: aws.ec2.RouteTableAssociation[];
+    for (const range of (await zones.then(zones => zones.names)).map((k, v) => {key: k, value: v})) {
+        rta.push(new aws.ec2.RouteTableAssociation(`rta-${range.key}`, {
+            routeTableId: eksRouteTable.id,
+            subnetId: vpcSubnet.then(vpcSubnet => vpcSubnet[range.key].id),
+        }));
+    }
+    const subnetIds = pulumi.all([pulumi.output(vpcSubnet), __item.id]).apply(([vpcSubnet, id]) => vpcSubnet.map(__item => id));
```
Comment (lblackstone) [inline pkg/codegen/internal/test/testdata/aws-eks.pp.ts:53]: Any reason `vpcSubnet` needs to be wrapped in `pulumi.output`?

Comment (pgavlin) [reply pkg/codegen/internal/test/testdata/aws-eks.pp.ts:53]: Hm, it doesn't, does it. Good catch--can you file a codegen issue for this? Something like "do not wrap non-outputs in `pulumi.output` when calling `pulumi.all`".

Comment (mikhailshilkov) [inline pkg/codegen/internal/test/testdata/aws-eks.pp.ts:53]: I don't quite understand this line at all... What is `__item` and why is it not
```js
const subnetIds = pulumi.all(vpcSubnet).apply(vpcSubnet => vpcSubnet.map(v => v.id));
```
?

Comment (pgavlin) [reply pkg/codegen/internal/test/testdata/aws-eks.pp.ts:53]: Hm, yeah, the generated code here does look wrong. This shouldn't involve applies at all, it should only involve `then`. Something like:

```typescript
const subnetIds = vpcSubnet.then(vpcSubnet => vpcSubnet.map(vpcSubnet => vpcSubnet.id));
```

Comment (pgavlin) [reply pkg/codegen/internal/test/testdata/aws-eks.pp.ts:53]: (`vpcSubnet` is a `Promise<aws.ec2.Subnet[]>`)

File: pkg/codegen/internal/test/testdata/aws-eks.pp.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            tags: {
+                Name: `pulumi-sn-${range.value}`,
+            },
+        }));
+    }
+    const rta: aws.ec2.RouteTableAssociation[];
+    for (const range of (await zones.then(zones => zones.names)).map((k, v) => {key: k, value: v})) {
+        rta.push(new aws.ec2.RouteTableAssociation(`rta-${range.key}`, {
+            routeTableId: eksRouteTable.id,
+            subnetId: vpcSubnet.then(vpcSubnet => vpcSubnet[range.key].id),
```
Comment (mikhailshilkov) [inline pkg/codegen/internal/test/testdata/aws-eks.pp.ts:50]: `vpcSubnet` is an array - so `then` looks wrong here?

File: pkg/codegen/internal/test/testdata/aws-eks.pp.ts (hunk 1 above)
Comment (mikhailshilkov) [reply pkg/codegen/internal/test/testdata/aws-eks.pp.ts:53]: Opened https://github.com/pulumi/pulumi/issues/4588

File: pkg/codegen/internal/test/testdata/aws-eks.pp.ts (hunk 2 above)
Comment (pgavlin) [reply pkg/codegen/internal/test/testdata/aws-eks.pp.ts:50]: `vpcSubnet` is a `Promise<aws.ec2.Subnet[]>`, so the `then` is correct.

Comment (mikhailshilkov) [reply pkg/codegen/internal/test/testdata/aws-eks.pp.ts:50]: It's declared as `const vpcSubnet: aws.ec2.Subnet[];` - why is it a promise?

Comment (pgavlin) [reply pkg/codegen/internal/test/testdata/aws-eks.pp.ts:50]: Ah, that's another bug :)

Comment (pgavlin) [reply pkg/codegen/internal/test/testdata/aws-eks.pp.ts:50]: Oh, hm. This is interesting. We should be retyping these for Node in a much deeper fashion.

**Pull request merged**